mostlyai/engine/_tabular/generation.py

```diff
@@ -924,6 +1118,17 @@ def generate(
             ctx_data, no_of_batches = _regroup_partial_sequences_by_length(
                 ctx_data, seed_data, ctx_primary_key, tgt_context_key
             )
+            # regroup by trailing NULL pattern for sequential imputation
+            if imputation and seed_data is not None and len(seed_data) > 0:
+                ctx_data, no_of_batches = _regroup_sequential_by_null_pattern(
+                    ctx_data, seed_data, ctx_primary_key, tgt_context_key, imputation.columns
+                )
+
+        # regroup by NULL pattern if imputation is enabled (flat data only)
+        if imputation and seed_data is not None and len(seed_data) > 0 and not is_sequential:
+            ctx_data, no_of_batches = _regroup_flat_by_null_pattern(
+                ctx_data, seed_data, ctx_primary_key, imputation.columns
+            )
```

Comment:
maybe just one block like this, but select `_regroup_sequential_by_null_pattern` or `_regroup_flat_by_null_pattern` based on `is_sequential` flag?